chain/pkg/bandrng/sampling.go

```diff
@@ -0,0 +1,40 @@
+package bandrng
+
+import "math"
+
+// AddUint64Overflow performs the addition operation on two uint64 integers and
+// returns a boolean on whether or not the result overflows.
+func AddUint64Overflow(a, b uint64) (uint64, bool) {
+	if math.MaxUint64-a < b {
+		return 0, true
+	}
+
+	return a + b, false
+}
+
+// SamplingOne sampling an index weighted by probability
+func SamplingOne(rng *Rng, weights []uint64) int {
```

Comment:
`// ChooseOne randomly picks an index between 0 and len(weights)-1 inclusively. Each index has the probability of getting selected based on its weight.`